File: android/Reguerta/app/src/main/java/com/reguerta/user/domain/startup/SemanticVersionComparator.kt
package com.reguerta.user.domain.startup

object SemanticVersionComparator {
    private val versionRegex = Regex("^\\d+(?:\\.\\d+)*$")

    fun compare(lhs: String, rhs: String): Int? {
        val leftParts = parseVersion(lhs) ?: return null
        val rightParts = parseVersion(rhs) ?: return null
        val maxSize = maxOf(leftParts.size, rightParts.size)

        repeat(maxSize) { index ->
            val left = leftParts.getOrElse(index) { 0 }
            val right = rightParts.getOrElse(index) { 0 }
            if (left != right) {
                return left.compareTo(right)
            }
        }

        return 0
    }

    private fun parseVersion(raw: String): List<Int>? {
        val value = raw.trim()
        if (value.isEmpty() || !versionRegex.matches(value)) {
            return null
        }

        return value.split(".").map { segment ->
            segment.toIntOrNull() ?: return null
        }
    }
}

File: ios/Reguerta/Reguerta/Domain/Startup/SemanticVersionComparator.swift
import Foundation

enum SemanticVersionComparator {
    private static let pattern = "^\\d+(?:\\.\\d+)*$"

    /// Compares numeric, dot-separated application versions.
    ///
    /// Surrounding whitespace is ignored and missing trailing components are treated as zero,
    /// so `1.2` and `1.2.0` compare as equal. Labels, empty components, negative values, and
    /// components that overflow `Int` are rejected.
    ///
    /// - Parameters:
    ///   - lhs: The version on the left side of the comparison.
    ///   - rhs: The version on the right side of the comparison.
    /// - Returns: `-1` when `lhs` is older, `0` when equal, `1` when newer, or `nil` when either
    ///   value is not a supported numeric version.
    static func compare(_ lhs: String, _ rhs: String) -> Int? {
        guard let leftParts = parse(lhs), let rightParts = parse(rhs) else {
            return nil
        }

        let maxCount = max(leftParts.count, rightParts.count)
        for index in 0..<maxCount {
            let left = index < leftParts.count ? leftParts[index] : 0
            let right = index < rightParts.count ? rightParts[index] : 0
            if left != right {
                return left < right ? -1 : 1
            }
        }

        return 0
    }

    private static func parse(_ raw: String) -> [Int]? {
        let value = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !value.isEmpty,
              value.range(of: pattern, options: .regularExpression) != nil
        else {
            return nil
        }

        var components: [Int] = []
        for component in value.split(separator: ".") {
            guard let parsed = Int(component) else {
                return nil
            }
            components.append(parsed)
        }
        return components
    }
}
